File: scripts/ingest_dipres.py
"""
ingest_dipres.py
Descarga y normaliza datos de ejecución presupuestaria de DIPRES para los
Servicios de Salud (Partida 16 - Ministerio de Salud).

Fuente: https://datos.gob.cl / https://www.dipres.gob.cl
Salida: data/dipres_ejecucion.json

Uso:
    pip install requests pandas openpyxl
    python scripts/ingest_dipres.py
"""

import json
import sys
from pathlib import Path
from datetime import datetime

ROOT = Path(__file__).parent.parent

SERVICIO_SALUD_IDS = {
    "SERVICIO DE SALUD METROPOLITANO NORTE": "ss_metro_norte",
    "SERVICIO DE SALUD METROPOLITANO OCCIDENTE": "ss_metro_occidente",
    "SERVICIO DE SALUD METROPOLITANO CENTRAL": "ss_metro_central",
    "SERVICIO DE SALUD METROPOLITANO ORIENTE": "ss_metro_oriente",
    "SERVICIO DE SALUD METROPOLITANO SUR": "ss_metro_sur",
    "SERVICIO DE SALUD METROPOLITANO SUR ORIENTE": "ss_metro_sur_oriente",
    "SERVICIO DE SALUD VALPARAISO - SAN ANTONIO": "ss_valparaiso_sa",
    "SERVICIO DE SALUD VINA DEL MAR - QUILLOTA": "ss_vina_quillota",
    "SERVICIO DE SALUD ACONCAGUA": "ss_aconcagua",
    "SERVICIO DE SALUD LIBERTADOR GENERAL BERNARDO O'HIGGINS": "ss_ohiggins",
    "SERVICIO DE SALUD MAULE": "ss_maule",
    "SERVICIO DE SALUD NUBLE": "ss_nuble",
    "SERVICIO DE SALUD BIOBIO": "ss_biobio",
    "SERVICIO DE SALUD TALCAHUANO": "ss_talcahuano",
    "SERVICIO DE SALUD ARAUCO": "ss_arauco",
    "SERVICIO DE SALUD ARAUCANIA NORTE": "ss_araucania_norte",
    "SERVICIO DE SALUD ARAUCANIA SUR": "ss_araucania_sur",
    "SERVICIO DE SALUD VALDIVIA": "ss_valdivia",
    "SERVICIO DE SALUD OSORNO": "ss_osorno",
    "SERVICIO DE SALUD DEL RELONCAVI": "ss_reloncavi",
    "SERVICIO DE SALUD CHILOE": "ss_chiloe",
    "SERVICIO DE SALUD AYSEN": "ss_aysen",
    "SERVICIO DE SALUD MAGALLANES": "ss_magallanes",
    "SERVICIO DE SALUD COQUIMBO": "ss_coquimbo",
    "SERVICIO DE SALUD ATACAMA": "ss_atacama",
    "SERVICIO DE SALUD ANTOFAGASTA": "ss_antofagasta",
    "SERVICIO DE SALUD TARAPACA": "ss_tarapaca",
    "SERVICIO DE SALUD ARICA Y PARINACOTA": "ss_arica_parinacota",
}

HERO_IDS = {"ss_metro_norte", "ss_araucania_sur", "ss_valparaiso_sa"}


def normalize_nombre(nombre: str) -> str:
    import unicodedata
    nombre = nombre.upper().strip()
    nombre = "".join(
        c for c in unicodedata.normalize("NFD", nombre)
        if unicodedata.category(c) != "Mn"
    )
    return nombre


def ingest_from_api():
    """
    Intenta obtener datos desde la API de datos.gob.cl / DIPRES.

    DIPRES publica archivos Excel mensuales con ejecución presupuestaria.
    URL típica: https://www.dipres.gob.cl/598/articles-XXXXX_doc_xls.xls

    También disponible en: https://datos.gob.cl/dataset/ejecucion-presupuestaria
    """
    try:
        import requests
        import pandas as pd

        print("Intentando conectar con API datos.gob.cl...")

        # Buscar datasets de DIPRES en datos.gob.cl
        url_api = "https://datos.gob.cl/api/3/action/package_search"
        params = {
            "q": "ejecucion presupuestaria dipres salud",
            "rows": 5,
        }
        r = requests.get(url_api, params=params, timeout=30)
        r.raise_for_status()

        results = r.json().get("result", {}).get("results", [])
        print(f"Datasets encontrados: {len(results)}")

        for ds in results:
            print(f"  - {ds.get('title', 'sin título')}: {ds.get('name', '')}")

        print("\nNOTA: La descarga e interpretación del Excel de DIPRES requiere")
        print("conocer la estructura exacta del archivo para el año consultado.")
        print("Revisar manualmente en: https://www.dipres.gob.cl/597/w3-propertyvalue-15131.html")

        return None

    except ImportError:
        print("ERROR: Instala dependencias: pip install requests pandas openpyxl")
        return None
    except Exception as e:
        print(f"ERROR al conectar con API: {e}")
        return None


def parse_excel_dipres(filepath: str, ano: int = 2024):
    """
    Parsea el Excel de ejecución presupuestaria de DIPRES.

    Estructura típica del Excel DIPRES:
    - Columna "Institución" o "Nombre": nombre del servicio
    - Columna "Presupuesto Inicial": monto asignado
    - Columna "Presupuesto Vigente": monto vigente (con modificaciones)
    - Columna "Gasto Devengado": monto efectivamente ejecutado

    Args:
        filepath: ruta al archivo Excel de DIPRES
        ano: año del presupuesto

    Returns:
        dict con datos normalizados por servicio_id
    """
    try:
        import pandas as pd

        print(f"Leyendo {filepath}...")
        df = pd.read_excel(filepath, header=None)

        print(f"Dimensiones: {df.shape}")
        print("Primeras filas:")
        print(df.head(10).to_string())

        print("\nNOTA: Ajusta las referencias de columnas según la estructura")
        print("real del archivo Excel descargado de DIPRES.")

        servicios = {}
        mes_corte = f"Diciembre {ano}"

        for _, row in df.iterrows():
            nombre_raw = str(row.get(0, "")).strip()
            nombre_norm = normalize_nombre(nombre_raw)

            if nombre_norm in SERVICIO_SALUD_IDS:
                sid = SERVICIO_SALUD_IDS[nombre_norm]
                try:
                    ppto_inicial = float(row.get(1, 0)) / 1_000_000
                    ppto_vigente = float(row.get(2, 0)) / 1_000_000
                    devengado = float(row.get(3, 0)) / 1_000_000
                    pct = (devengado / ppto_vigente * 100) if ppto_vigente > 0 else 0

                    servicios[sid] = {
                        "presupuesto_inicial_MM": round(ppto_inicial, 0),
                        "presupuesto_vigente_MM": round(ppto_vigente, 0),
                        "devengado_MM": round(devengado, 0),
                        "pct_ejecucion": round(pct, 1),
                        "mes_corte": mes_corte,
                        "es_hero": sid in HERO_IDS,
                    }
                except (ValueError, TypeError) as e:
                    print(f"  WARN: Error parseando {nombre_raw}: {e}")

        print(f"Servicios encontrados: {len(servicios)}")
        return servicios

    except ImportError:
        print("ERROR: pip install pandas openpyxl")
        return None
    except Exception as e:
        print(f"ERROR: {e}")
        return None


def save_output(servicios: dict, ano: int = 2024, mes_corte: str = "Diciembre 2024"):
    """Guarda el JSON de salida."""
    existing_path = ROOT / "data" / "dipres_ejecucion.json"

    with open(existing_path) as f:
        existing = json.load(f)

    for sid, data in servicios.items():
        if sid in existing["servicios"]:
            existing["servicios"][sid].update(data)

    existing["ano"] = ano
    existing["mes_corte"] = mes_corte
    existing["fecha_actualizacion"] = datetime.now().strftime("%Y-%m-%d")

    with open(existing_path, "w", encoding="utf-8") as f:
        json.dump(existing, f, ensure_ascii=False, indent=2)

    print(f"\nGuardado en {existing_path}")
    print(f"Servicios actualizados: {len(servicios)}")


if __name__ == "__main__":
    print("=" * 60)
    print("Ingesta DIPRES - Ejecución Presupuestaria Salud")
    print("=" * 60)

    if len(sys.argv) > 1:
        excel_path = sys.argv[1]
        ano = int(sys.argv[2]) if len(sys.argv) > 2 else 2024
        print(f"\nModo Excel: {excel_path} (año {ano})")
        servicios = parse_excel_dipres(excel_path, ano)
        if servicios:
            save_output(servicios, ano)
    else:
        print("\nModo exploración (sin archivo Excel)")
        print("Uso: python scripts/ingest_dipres.py <ruta_excel.xlsx> [año]")
        print("\nExplorando API datos.gob.cl...")
        ingest_from_api()
        print("\nPara actualizar datos:")
        print("1. Descarga el Excel de DIPRES desde:")
        print("   https://www.dipres.gob.cl/597/w3-propertyvalue-15131.html")
        print("2. Ejecuta: python scripts/ingest_dipres.py ruta_al_archivo.xlsx 2024")

File: scripts/ingest_dipres.mjs
#!/usr/bin/env node
/**
 * scripts/ingest_dipres.mjs
 *
 * Reescribe data/dipres_ejecucion.json descargando los XMLs oficiales de
 * ejecución presupuestaria DIPRES por Servicio de Salud (Partida 16, Salud).
 * Cero dependencias: solo Node >= 18.
 *
 * Uso:   node scripts/ingest_dipres.mjs
 * o:     npm run ingest:dipres
 *
 * Regla CLAUDE.md #1: cifras null si la fuente no responde o no contiene
 * el bloque GASTOS. Cada registro lleva url_fuente_real.
 */

import { writeFileSync, readFileSync } from 'node:fs'
import { resolve, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const ROOT = resolve(__dirname, '..')
const OUT_PATH = resolve(ROOT, 'data/dipres_ejecucion.json')

// Mapeo ss_id → article id DIPRES (Servicios de Salud, 28 en total).
// Estos IDs son los publicados por DIPRES para los informes trimestrales
// de ejecución. Si DIPRES cambia el ID, actualizar esta tabla.
const ARTICLE_IDS = {
  ss_arica_parinacota:  '359477',
  ss_tarapaca:          '359478',
  ss_antofagasta:       '359479',
  ss_atacama:           '359480',
  ss_coquimbo:          '359481',
  ss_valparaiso_sa:     '359482',
  ss_vina_quillota:     '359483',
  ss_aconcagua:         '359484',
  ss_ohiggins:          '359485',
  ss_maule:             '359486',
  ss_nuble:             '359487',
  ss_talcahuano:        '359489',
  ss_biobio:            '359490',
  ss_arauco:            '359491',
  ss_araucania_norte:   '359492',
  ss_araucania_sur:     '359493',
  ss_osorno:            '359494',
  ss_valdivia:          '359495', // XML real = "SERVICIO DE SALUD LOS RÍOS"
  ss_reloncavi:         '359496',
  ss_aysen:             '359497',
  ss_magallanes:        '359498',
  ss_metro_oriente:     '359499',
  ss_metro_central:     '359500',
  ss_metro_sur:         '359501',
  ss_metro_norte:       '359502',
  ss_metro_occidente:   '359503',
  ss_metro_sur_oriente: '359504',
  ss_chiloe:            '359508',
}

const HERO_IDS = new Set(['ss_metro_norte', 'ss_araucania_sur', 'ss_valparaiso_sa'])

const XML_BASE = 'https://www.dipres.gob.cl/597/articles-'

const round2 = (n) => (n == null ? null : Math.round(n * 100) / 100)

function parseGastos(xml) {
  // Captura el bloque <nombre>GASTOS</nombre> seguido de los 3 montos.
  const m = xml.match(
    /<nombre>GASTOS<\/nombre>\s*<formulado>(\d+)<\/formulado>\s*<vigente>(\d+)<\/vigente>\s*<monto>(\d+)<\/monto>/
  )
  if (!m) return null
  return {
    formulado_raw: parseInt(m[1], 10),
    vigente_raw: parseInt(m[2], 10),
    monto_raw: parseInt(m[3], 10),
  }
}

function parseNombreServicio(xml) {
  const m = xml.match(/<nombre>(SERVICIO DE SALUD[^<]+)<\/nombre>/)
  return m ? m[1].trim() : null
}

async function fetchService(key, articleId) {
  const url = XML_BASE + articleId + '_doc_xml.xml'
  const r = await fetch(url)
  if (r.status !== 200) {
    return { key, url, ok: false, error: `HTTP ${r.status}` }
  }
  const xml = await r.text()
  const gastos = parseGastos(xml)
  const nombre_oficial = parseNombreServicio(xml)
  if (!gastos) return { key, url, ok: false, error: 'No <GASTOS> en XML' }
  return { key, url, ok: true, gastos, nombre_oficial }
}

async function main() {
  const existing = JSON.parse(readFileSync(OUT_PATH, 'utf8'))
  const oldServicios = existing.servicios ?? {}

  console.log('Descargando', Object.keys(ARTICLE_IDS).length, 'XMLs oficiales de dipres.gob.cl ...')
  const results = await Promise.all(
    Object.entries(ARTICLE_IDS).map(([k, a]) => fetchService(k, a).catch((e) => ({ key: k, ok: false, error: e.message })))
  )

  const servicios = {}
  let nFail = 0
  for (const r of results) {
    const old = oldServicios[r.key] ?? {}
    if (!r.ok) {
      nFail++
      console.warn('  FAIL', r.key, r.error)
      servicios[r.key] = {
        nombre: old.nombre ?? r.key,
        region: old.region ?? null,
        presupuesto_inicial_MM: null,
        presupuesto_vigente_MM: null,
        devengado_MM: null,
        pct_ejecucion: null,
        poblacion_beneficiaria_aprox: null,
        mes_corte: null,
        url_fuente_real: r.url ?? null,
        es_hero: HERO_IDS.has(r.key),
        nota: `Error en ingesta: ${r.error}`,
      }
      continue
    }
    const g = r.gastos
    const vigenteMM = g.vigente_raw / 1000
    const monto = g.monto_raw / 1000
    const pct = vigenteMM > 0 ? (monto / vigenteMM) * 100 : null
    servicios[r.key] = {
      nombre: old.nombre ?? (r.nombre_oficial
        ? r.nombre_oficial.replace(/^SERVICIO DE SALUD/, 'Servicio de Salud')
        : r.key),
      region: old.region ?? null,
      presupuesto_inicial_MM: round2(g.formulado_raw / 1000),
      presupuesto_vigente_MM: round2(vigenteMM),
      devengado_MM: round2(monto),
      pct_ejecucion: round2(pct),
      poblacion_beneficiaria_aprox: null,
      mes_corte: old.mes_corte ?? 'Cuarto trimestre 2024',
      url_fuente_real: r.url,
      es_hero: HERO_IDS.has(r.key),
    }
  }

  const out = {
    ...existing,
    version: '1.2.0',
    fecha_ingesta: new Date().toISOString(),
    servicios,
  }

  writeFileSync(OUT_PATH, JSON.stringify(out, null, 2))
  console.log('Escrito', OUT_PATH)
  console.log('  Servicios OK:', Object.keys(servicios).length - nFail, '/ Fallidos:', nFail)
}

main().catch((e) => {
  console.error('FATAL:', e)
  process.exit(1)
})
